recipes/TIMIT/ASR/transducer/train_wav2vec.py

```diff
@@ -0,0 +1,376 @@
+        return p_transducer
+
+    def compute_objectives(self, predictions, batch, stage):
+        "Given the network predictions and targets computed the loss."
+        ids = batch.id
+        _, wav_lens = batch.sig
+        phns, phn_lens = batch.phn_encoded
+        if stage != sb.Stage.TRAIN:
+            predictions, hyps = predictions
+
+        phns = phns.long()
+        loss = self.hparams.compute_cost(predictions, phns, wav_lens, phn_lens)
+        self.transducer_metrics.append(
+            ids, predictions, phns, wav_lens, phn_lens
+        )
+
+        if stage != sb.Stage.TRAIN:
+            self.per_metrics.append(
+                ids, hyps, phns, None, phn_lens, self.label_encoder.decode_ndim
+            )
+
+        return loss
+
+    def on_stage_start(self, stage, epoch):
+        "Gets called when a stage (either training, validation, test) starts."
+        self.transducer_metrics = self.hparams.transducer_stats()
+
+        if stage != sb.Stage.TRAIN:
+            self.per_metrics = self.hparams.per_stats()
+
+    def on_stage_end(self, stage, stage_loss, epoch):
+        """Gets called at the end of a epoch."""
+        if stage == sb.Stage.TRAIN:
+            self.train_loss = stage_loss
+        else:
+            per = self.per_metrics.summarize("error_rate")
+
+        if stage == sb.Stage.VALID:
+            old_lr_adam, new_lr_adam = self.hparams.lr_annealing_adam(per)
+            old_lr_wav2vec, new_lr_wav2vec = self.hparams.lr_annealing_wav2vec(
+                per
+            )
+            sb.nnet.schedulers.update_learning_rate(
+                self.adam_optimizer, new_lr_adam
+            )
+            sb.nnet.schedulers.update_learning_rate(
+                self.wav2vec_optimizer, new_lr_wav2vec
+            )
+
+            self.hparams.train_logger.log_stats(
+                stats_meta={
+                    "epoch": epoch,
+                    "lr_adam": old_lr_adam,
+                    "lr_wav2vec": old_lr_wav2vec,
+                },
+                train_stats={"loss": self.train_loss},
+                valid_stats={"loss": stage_loss, "PER": per},
+            )
+            self.checkpointer.save_and_keep_only(
+                meta={"PER": per}, min_keys=["PER"]
+            )
+
+        if stage == sb.Stage.TEST:
+            self.hparams.train_logger.log_stats(
+                stats_meta={"Epoch loaded": self.hparams.epoch_counter.current},
+                test_stats={"loss": stage_loss, "PER": per},
+            )
+            with open(self.hparams.wer_file, "w") as w:
+                w.write("Transducer loss stats:\n")
+                self.transducer_metrics.write_stats(w)
+                w.write("\nPER stats:\n")
+                self.per_metrics.write_stats(w)
+                print(
+                    "Transducer and PER stats written to file",
+                    self.hparams.wer_file,
+                )
+
+    def fit_batch(self, batch):
```

Comment:
Remove the docstring and add instead some comments on why we need to override it (dual optimizer)

Reply:
I use the same docstring as `seq2seq/train_wav2vec.py` which explain that fit_batch override to do multiple updates